catcher_modules/mq/rabbit.py

```diff
@@ -0,0 +1,164 @@
+                data: '{{ data|tojson }}'
+
+    Publish `data_from_file` variable as json message
+    ::
+        variables:
+            rabbitmq_config:
+                url: 127.0.0.1:5672
+                username: 'guest'
+                password: 'guest'
+        rabbit:
+            publish:
+                config: '{{ rabbitmq_config }}''
+                exchange: 'test.catcher.exchange'
+                routing_key: 'catcher.routing.key'
+                data_from_file: '{{ /path/to/file }}'            
+    """
+
+    @update_variables
+    def action(self, includes: dict, variables: dict) -> any:
+        body = self.simple_input(variables)
+        self.method = Step.filter_predefined_keys(body)  # publish/consume
+
+        operation = body[self.method]
+
+        #if virtual host is not specified default it to /
+        config = operation['config']
+        if config.get('virtualhost') is None:
+            config['virtualhost'] = ''
+
+        connectionParameters = self._get_connection_parameters(config)
+
+        if self.method == 'publish':
+            message = self._get_data(operation)
+            return variables, self.publish(connectionParameters, operation['exchange'], operation['routing_key'], operation.get('headers'), message)
+        elif self.method == 'consume':
+            return variables, self.consume(connectionParameters, operation['queue'])    
+        else:
+            raise AttributeError('unknown method: ' + self.method)
+        
+    def publish(self, connectionParameters, exchange, routingKey, headers, message):
+        import pika
+        properties = pika.BasicProperties(headers=headers)
+        with pika.BlockingConnection(connectionParameters) as connection:
+            channel = connection.channel()
+            channel.basic_publish(exchange=exchange,routing_key=routingKey,properties=properties,body=message)
+
+    def consume(self, connectionParameters, queue):
+        message = None
+        import pika
+        with pika.BlockingConnection(connectionParameters) as connection:
+            channel = connection.channel()
+            method_frame, header_frame, body = channel.basic_get(queue)
+            if method_frame:
+                channel.basic_ack(method_frame.delivery_tag)
+                message = try_get_object(body.decode('UTF-8'))
+        return message
+    
+    def _get_data(self, operation):
+        if operation.get('data') is not None:
+            return str(operation.get('data'))
+        elif operation.get('data_from_file') is not None:
+            with open(operation['data_from_file'], 'r') as f:
+                return f.read()
+        raise AttributeError('data or data_from_file should be passed: ' + self.method)
+
+    def _get_connection_parameters(self, config):
+        import pika
+        amqpURL = 'amqp://{}:{}@{}/{}'
+        amqpsURL = 'amqps://{}:{}@{}/{}'
+        sslOptions = config.get('sslOptions')
+        if sslOptions is not None:
```

Comment:
Would be better to set up `s` as the other arguments and form URLParameters only once.
```
amqpURL = 'amqp{}://{}:{}@{}/{}'
sslOptions = config.get('sslOptions')
parameters = pika.URLParameters(amqpsURL.format('s' if sslOptions else '', config['username'], config['password'], config['server'], config['virtualhost']))
if sslOptions is not None:
    parameters.ssl = True
    parameters.ssl_options = self._get_ssl_options(sslOptions)
return parameters
```